OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs

```diff
@@ -0,0 +1,357 @@
+				return;
+
+			isEnemyUnit = unit =>
+				Player.Stances[unit.Owner] == Stance.Enemy
+					&& !unit.Info.HasTraitInfo<HuskInfo>()
+					&& unit.Info.HasTraitInfo<ITargetableInfo>();
+
+			unitCannotBeOrdered = a => a.Owner != Player || a.IsDead || !a.IsInWorld;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			notifyPositionsUpdated = Player.PlayerActor.TraitsImplementing<IBotPositionsUpdated>().ToArray();
+			notifyIdleBaseUnitCount = Player.PlayerActor.TraitsImplementing<IBotIdleBaseUnits>().ToArray();
+
+			// Avoid all AIs trying to rush in the same tick, randomize their initial rush a little.
+			var smallFractionOfRushInterval = Info.RushInterval / 20;
+			rushTicks = World.LocalRandom.Next(Info.RushInterval - smallFractionOfRushInterval, Info.RushInterval + smallFractionOfRushInterval);
+
+			// Avoid all AIs reevaluating assignments on the same tick, randomize their initial evaluation delay.
+			assignRolesTicks = World.LocalRandom.Next(0, Info.AssignRolesInterval);
+			attackForceTicks = World.LocalRandom.Next(0, Info.AttackForceInterval);
+			minAttackForceDelayTicks = World.LocalRandom.Next(0, Info.MinimumAttackForceDelay);
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			AssignRolesToIdleUnits(bot);
+		}
+
+		internal Actor FindClosestEnemy(WPos pos)
+		{
+			return World.Actors.Where(isEnemyUnit).ClosestTo(pos);
+		}
+
+		internal Actor FindClosestEnemy(WPos pos, WDist radius)
+		{
+			return World.FindActorsInCircle(pos, radius).Where(isEnemyUnit).ClosestTo(pos);
+		}
+
+		void CleanSquads()
+		{
+			Squads.RemoveAll(s => !s.IsValid);
+			foreach (var s in Squads)
+				s.Units.RemoveAll(unitCannotBeOrdered);
+		}
+
+		// Use of this function requires that there is one squad of this type. Hence it is a piece of shit
+		Squad GetSquadOfType(SquadType type)
+		{
+			return Squads.FirstOrDefault(s => s.Type == type);
+		}
+
+		Squad RegisterNewSquad(IBot bot, SquadType type, Actor target = null)
+		{
+			var ret = new Squad(bot, this, type, target);
+			Squads.Add(ret);
+			return ret;
+		}
+
+		void AssignRolesToIdleUnits(IBot bot)
+		{
+			CleanSquads();
+
+			activeUnits.RemoveAll(unitCannotBeOrdered);
+			unitsHangingAroundTheBase.RemoveAll(unitCannotBeOrdered);
+			foreach (var n in notifyIdleBaseUnitCount)
+				n.UpdatedUnitsHangingAroundBaseCount(unitsHangingAroundTheBase.Count);
+
+			if (--rushTicks <= 0)
+			{
+				rushTicks = Info.RushInterval;
+				TryToRushAttack(bot);
+			}
+
+			if (--attackForceTicks <= 0)
+			{
+				attackForceTicks = Info.AttackForceInterval;
+				foreach (var s in Squads)
+					s.Update();
+			}
+
+			if (--assignRolesTicks <= 0)
+			{
+				assignRolesTicks = Info.AssignRolesInterval;
+				FindNewUnits(bot);
+			}
+
+			if (--minAttackForceDelayTicks <= 0)
+			{
+				minAttackForceDelayTicks = Info.MinimumAttackForceDelay;
+				CreateAttackForce(bot);
+			}
+		}
+
+		void FindNewUnits(IBot bot)
+		{
+			var newUnits = World.ActorsHavingTrait<IPositionable>()
```

Comment:
General thought for future improvements: This will pick up all trees/rocks/civ buildings on the map. If we can find a good excuse to introduce some kind of `ControllableByBot` trait with actor-specific data then we could filter on that here instead.

Reply:
But chances are that this trait would be introduced as part of a proper replacement for `SquadManagerBotModule`, so ¯\\_(ツ)_/¯

Reply:
I think it's unrealistic for mods to have MCVs that don't implement `IMove` somewhere, so I'll go for that instead.
Modders that really don't use any trait implementing `IMove` for MCVs should be able to write/customize their own McvManager module, too.